dhcp6/random_address_pool.go

```diff
@@ -0,0 +1,168 @@
+	ret := &RandomAddressPool{}
+	ret.validLifetime = validLifetime
+	ret.poolStartAddress = big.NewInt(0)
+	ret.poolStartAddress.SetBytes(poolStartAddress)
+	ret.poolSize = poolSize
+	ret.identityAssociations = make(map[uint64]*IdentityAssociation)
+	ret.usedIps = make(map[uint64]struct{})
+	ret.identityAssociationExpirations = newFifo()
+	ret.timeNow = func() time.Time { return time.Now() }
+
+	ticker := time.NewTicker(time.Second * 10).C
+	go func() {
+		for {
+			<- ticker
+			ret.ExpireIdentityAssociations()
+		}
+	}()
+
+	return ret
+}
+
+// ReserveAddresses creates new or retrieves active associations for interfaces in interfaceIDs list.
+func (p *RandomAddressPool) ReserveAddresses(clientID []byte, interfaceIDs [][]byte) ([]*IdentityAssociation, error) {
+	p.lock.Lock()
+	defer p.lock.Unlock()
+
+	ret := make([]*IdentityAssociation, 0, len(interfaceIDs))
+	rng := rand.New(rand.NewSource(p.timeNow().UnixNano()))
+
+	for _, interfaceID := range (interfaceIDs) {
+		clientIDHash := p.calculateIAIDHash(clientID, interfaceID)
+		association, exists := p.identityAssociations[clientIDHash];
+
+		if exists {
+			ret = append(ret, association)
+			continue
+		}
+		if uint64(len(p.usedIps)) == p.poolSize {
+			return ret, fmt.Errorf("No more free ip addresses are currently available in the pool")
+		}
+
+		for {
+			// we assume that ip addresses adhere to high 64 bits for net and subnet ids, low 64 bits are for host id rule
+			hostOffset := randomUint64(rng) % p.poolSize
+			newIP := big.NewInt(0).Add(p.poolStartAddress, big.NewInt(0).SetUint64(hostOffset))
+			_, exists := p.usedIps[newIP.Uint64()];
+			if !exists {
+				timeNow := p.timeNow()
+				association := &IdentityAssociation{ClientID: clientID,
+					InterfaceID: interfaceID,
+					IPAddress: newIP.Bytes(),
+					CreatedAt: timeNow}
+				p.identityAssociations[clientIDHash] = association
+				p.usedIps[newIP.Uint64()] = struct{}{}
+				p.identityAssociationExpirations.Push(&associationExpiration{expiresAt: p.calculateAssociationExpiration(timeNow), ia: association})
+				ret = append(ret, association)
+				break
+			}
+		}
+	}
+
+	return ret, nil
+}
+
+// ReleaseAddresses returns IP addresses associated with ClientID and interfaceIDs back into the address pool
+func  (p *RandomAddressPool) ReleaseAddresses(clientID []byte, interfaceIDs [][]byte) {
+	p.lock.Lock()
+	defer p.lock.Unlock()
+
+	for _, interfaceID := range(interfaceIDs) {
+		association, exists := p.identityAssociations[p.calculateIAIDHash(clientID, interfaceID)]
+		if !exists {
+			continue
+		}
+		delete(p.usedIps, big.NewInt(0).SetBytes(association.IPAddress).Uint64())
+		delete(p.identityAssociations, p.calculateIAIDHash(clientID, interfaceID))
+	}
+}
+
+// ExpireIdentityAssociations releases IP addresses in identity associations that reached the end of valid lifetime
+// back into the address pool
+func (p *RandomAddressPool) ExpireIdentityAssociations() {
```

Comment:
I feel it is small enough not to cause any cognitive overload for anybody and I don't feel it will be expensive to maintain either. I'll remove it if you feel strongly about it though...